### ISSUE TITLE
Revisit Extraction Pipeline Design

### ISSUE THREAD

**Issue by ppanopticon** [vitrivr/vitrivr-engine#51]
# Task Description
The way extraction pipelines can be designed at the moment offers much flexibility. However, there still are certain cases that cannot be modeled due to some structural constraints imposed by how operators can be combined.  Given the recent insights, we propose the following design changes:

- [x] There are only three types of extraction `Operators`: `Enumerators`, `Decoders`, `Retrievers` and regular `Operators`. We do no longer distinguish between `Aggregator`, `Transformer` and `Exporter` at an interface level.
- [x] An `Enumerator` is strictly followed by a `Decoder`. After that, `Operators` can be freely mixed and matched. It remains at the discretion of an `Operator`, whether it creates a 1:1 or a 1:n mapping in terms of output `Retrievable`s.
- [x] All operators can be configured using named key-value pairs (naming of the keys is at the discretion of the implementing class).
- [x] Whether or not an operator produces a persistent `Descriptor`, `Retrievable` or `Relationship` can be specified using the reserved `transient` configuration key.
- [x] Additional global keys may be defined.

## Dependencies
- Interdependency with #27 

## Boundary Conditions
- Basically, this change relaxes constraints on how `Operators` can be combined. The issue should not go beyond that.
- The should not break with the existing definition language (until a replacement has been built).

**Comment by sauterl** [comment]
@ppanopticon I guess there are some discussions to be held regarding the details: You specifically wrote `Retriever` as a possible operator type to be mixed and matched. Currently, this is an `Exporter`. Is this a consequence to enable free mix'n'match, such that these operators operate on `Retrievable`s or are there other design concepts behind this?

**Comment by sauterl** [comment]
I suggest to first rework the definition language (#27), then tackle the relaxations drafted here and in doing so, adjust the definition language where necessary. Partly since I already started with the extraction pipieline definition and partly in order to not inforce a backwards compatibliity that adds an uncessecary layer of compelixity to the task.

**Comment by ppanopticon** [comment]
I think basically it comes down to allowing all `Operators` that have `Retrievables` as in- and output to be mixed and matched freely, regardless of what exactly it is they actually do (creating exports, extracting a feature etc.). 

Consequently, we no longer need an explicit distinction between `Aggregator`, `Transformer` and `Exporter`, since they all implement the same interface. `Extractors` are a bit different, since they have an explicit connection to a `field`. Nevertheless, by the same argument, they should also be mixable with other operators.

And last but not least, I would argue that `Decoder` should no longer generate `ContentElements` but instead `Retrievables` as well. Applying these principles, we end-up with three types of operators (from a processing perspective):

- 1:n mappers (what we refer to as `Segmenter`)
- 1:1 mappers (what we refer to as `Transformer`, `Extractor` and `Exporter`)
- n:1 mappers (what we refer to as `Aggergator`)

Of course the same logic applies to the query side of things.

**Comment by sauterl** [comment]
@ppanopticon Thanks for the clarification, this is what I have suspected.

For transparency I'd like to keep the discussion regarding the extration pipeline here:

@lucaro @ppanopticon @net-cscience-raphael Currently, an `Exporter` can be defined on the `Schema`, what is the reasoning behind this decision?
The consequences are, that during the redesign, the 1:1 mappers have more semantics on them:
* `Extractor`s have to be backed by a `Schema.Field`
* `Exporter`s _might_ be defined on the `Schema`

**Comment by faberf** [comment]
> @lucaro @ppanopticon @net-cscience-raphael Currently, an `Exporter` can be defined on the `Schema`, what is the reasoning behind this decision? 

If I remember correctly, it was to facilitate serving media objects by the server the same way cineast did previously. If the exporter was only defined in an extraction config these exported objects might be difficult to access later.

**Comment by sauterl** [comment]
@faberf  Thanks for the reply, in the meantime, the collegues answered this question off-line.
